Stats: Add highlights to Ads page

a8ck3n opened the pull request:
#### Proposed Changes

Add highlights section to top of Ads page.

The first pass at getting the highlights working. Uses the existing data from the store and formats it for display.

- Adds new highlights section at the top of the page
- Tooltip should show correct payout/status notices
- Hides old Totals section (below chart)

![AdsHighlightsSnap](https://user-images.githubusercontent.com/40267301/213646405-8f31e0b1-a66d-4f0b-8c23-d8acd759b9e6.png)

Does not yet:

* use correct icon data
* remove the previous implementation

I'd like to address these in follow up PRs.

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Visit the Calypso Live link.
* View a site with WordAds enabled.
* Visit Stats → Ads and confirm highlights are shown correctly.
* Confirm things work correctly on smaller screen sizes.
* Confirm hover works correctly with the tooltip.

If you need to compare with the old Totals section, you can enable it via the React dev tools. Select the `WordAdsEarnings` component and toggle the `showTotalsSection` property.

![DevTools](https://user-images.githubusercontent.com/40267301/213647731-a290ceb6-71e3-4267-a0ac-348321b28630.png)

The new code is only active on non-production domains so it can safely be merged as is.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #72045.

a8ck3n committed:
First pass at WordAds highlights section.

a8ck3n committed:
Add simplified HighlightCard component.

Logic-free component that accepts strings values.

a8ck3n committed:
Add some example values.

a8ck3n committed:
Add some environment checks.

a8ck3n committed:
Move section header to sub-component.

a8ck3n committed:
Refactor handling of card list.

a8ck3n committed:
Comments++

a8ck3n committed:
Factor out data handling.

a8ck3n committed:
Index and key data for React.

a8ck3n committed:
Pull data from store.

a8ck3n committed:
Properly format zero dollar amounts.

Watch for and replace “$0.00” with “$0”.

a8ck3n committed:
Increase min-width for these highlights.

The previous min-width of 180px was too narrow for both the “Outstanding amount” heading and any sums 5 figures or larger. New value of 280px is good for 7 figures plus decimals. Future proof?!

a8ck3n committed:
Update heading to match design.

a8ck3n committed:
Add context for translated strings.

a8ck3n committed:
Implement the info icon with hover state.

a8ck3n committed:
Tweak styles for info icon.

a8ck3n committed:
Fix for layout issue on smaller screens.

a8ck3n committed:
Hide the old Totals section.

Controlled by a property on the component for now. When the new highlights section is visible, the old Totals section is hidden. Can be toggled via the React dev tools.

a8ck3n committed:
Use comment instead of context for translation.

Per the header documentation, the comment should be used for translation hints.

a8ck3n committed:
Enable new highlight section more widely.

wilmersondasilva commented:
I noticed the `font-family` for the amounts is different from what we're using on the other Traffic and Insights pages and also from the screenshots on https://github.com/Automattic/wp-calypso/issues/72045. Are they different on purpose?

![image](https://user-images.githubusercontent.com/6406071/213820928-e399e69c-f7f0-41c3-b497-82d2add621bb.png)

wilmersondasilva reviewed (approved):
Aside from the `font-family` difference I mentioned above, it worked properly, and the code is fine too.

grzegorz-cp commented on client/my-sites/stats/wordads/index.jsx:
```diff
@@ -136,6 +137,10 @@ class WordAds extends Component {
 		);
 	};
 
+	isProduction() {
```
Could we use feature flags instead of a URL check? It could be done in a similar way to [this](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/stats/stats-insights/index.jsx#L31).

a8ck3n replied:
I purposefully avoided a feature flag in this case as this is a very localized change. If I didn't anticipate a follow up PR for the icons, I wouldn't need a check at all. Now that I've added in the icons, I think we can just ship this as production-ready though.

I'll make those changes first, then revisit this.

a8ck3n replied:
This check has been removed now!

grzegorz-cp commented on client/my-sites/stats/wordads/highlights-section.jsx:
```diff
@@ -0,0 +1,153 @@
+	return (
+		<h1 className="highlight-cards-heading">
+			{ localizedTitle }{ ' ' }
+			<Gridicon
```
Let's try to use `<Icon />` instead of a `<Gridicon />`. We are migrating as many icons as possible to help with loading them for Odyssey Stats.

a8ck3n replied:
Thanks for the heads up. I wasn't aware of this. I'll get this fixed up!

grzegorz-cp commented on client/my-sites/stats/wordads/highlights-section.jsx:
```diff
@@ -0,0 +1,153 @@
+	// Takes a Number, formats it to 2 decimal places, and prepends a "$".
+	// This mimics the existing behaviour. I'm assuming we only view/report
+	// on earnings in USD.
+	const formattedAmount = '$' + numberFormat( amount, 2 );
```
If we need a new number formatter, would you be able to [extend those](https://github.com/Automattic/wp-calypso/tree/trunk/packages/components/src/number-formatters) so we'd have them centralised?

a8ck3n replied:
I'm unsure about this suggestion. The `numberFormat()` used here is part of `i18n-calypso` (the same place our `translate()` call lives) and it's used extensively around Calypso. Does it make sense to introduce a new version of this or maybe to move our new formatters into `i18n-calypso`?

I'd lean towards using the existing version as it will safely replicate the behaviour of the code it's replacing.

Let me know if I'm missing something here that makes the number-formatters component the better choice though. I'm happy to move it if that's the better option.

grzegorz-cp replied:
That's a very good point. From what I can see, `numberFormat` with two parameters only formats a number but doesn't take locale into account, which means that for countries using a comma instead of a period before fractions, this won't be swapped. On the other hand, we hard-coded the dollar sign. Can this page show a different currency, or is it always USD?
Does showing a dollar value mean we use the US number format?
All the other highlight sections take locale into account (they use browser locale), so I lean towards using a localised number even with a foreign currency.
The table values make the same mistake, so if we swap, this wouldn't be constant until table numbers are also addressed.
This is not a blocking issue, so I can accept both outcomes, but we should probably create a task to look at this if we don't do it here.

a8ck3n replied:
My assumption was that WordAds only deals in USD amounts and my initial focus was to replicate the existing behaviour so as not to confuse things. The "$" symbol is hard-coded in the previous implementation as well.

My preference would be to address changes in the currency handling in a separate PR though, once we've had a chance to investigate and confirm that results should always be shown in USD (or not).

I'll spin up an issue to track this and I'll make a note that the behaviour should be consistent between the highlight cards and the values in the tables.

a8ck3n replied:
Opened an issues for this [here](https://github.com/Automattic/wp-calypso/issues/72492).

grzegorz-cp replied:
Great, thanks @a8ck3n.

grzegorz-cp commented on client/my-sites/stats/wordads/index.jsx:
```diff
@@ -245,7 +256,7 @@ class WordAds extends Component {
 
 								<div className="stats__module-list stats__module-headerless--unified">
-									<WordAdsEarnings site={ site } />
+									<WordAdsEarnings site={ site } showTotalsSection={ ! showNewHighlights } />
```
If you add a condition around this component, you won't have to pass the property. For example:
```
{ !isFeatureFlagEnabled && <WordAdsEarnings site={ site } /> }
```
It might also be better to add the condition around the whole `div` if it doesn't break the layout.

a8ck3n replied:
We can't conditionally render the WordAdsEarnings component because it also handles the render of up to three tables in addition to the totals section. If we used a feature flag we could do that check inside the component but, as noted above, I can remove this completely now that the new icons are part of the PR (incoming).

a8ck3n replied:
The old code path has been removed so the property is no longer needed.

grzegorz-cp commented on client/my-sites/stats/wordads/highlight-card-simple.jsx:
```diff
@@ -0,0 +1,14 @@
+import { Card } from '@automattic/components';
+
+function HighlightCardSimple( { heading, icon, value } ) {
```
For another variant of a highlight card, could you place it along with others [here](https://github.com/Automattic/wp-calypso/tree/trunk/packages/components/src/highlight-cards)? This way, we have all cards centralised making it easier if we decide to unify them or combine them somehow. We could also add a storybook story for it.

a8ck3n replied:
My thinking was that I'd open up a maintenance task to refactor the various highlight cards and then move them into the component. We also have a variant here:

https://github.com/Automattic/wp-calypso/tree/trunk/client/my-sites/stats/stats-email-top-row

The new HighlightCardSimple component and the TopCard component are mostly logic free so they should factor together well. Any objection to addressing this in a follow-up PR?

grzegorz-cp replied:
Sounds good. It will be great to refactor the cards as a maintenance task.

a8ck3n replied:
Opened a task for this [here](https://github.com/Automattic/wp-calypso/issues/72493).

grzegorz-cp commented on client/my-sites/stats/wordads/highlights-section.jsx:
```diff
@@ -0,0 +1,153 @@
+	);
+}
+
+export default function HighlightsSection( props ) {
```
Similar to the cards, would it be possible to put another variant of the header [here](https://github.com/Automattic/wp-calypso/tree/trunk/packages/components/src/highlight-cards)?

a8ck3n replied:
I hadn't considered this. I was thinking of the Storybook components as components that might be reusable outside Stats. For example, I can see other teams needing to make use of simple highlight cards. The section components are a bit more data-coupled though, so it made sense they would live inside Stats.

Are we thinking we should develop entire sections or layouts inside the components package, even if we don't see them as reusable?

grzegorz-cp replied:
I see your point now, and this file has many functions specific to ads. So let's keep it here.

grzegorz-cp reviewed (changes requested):
Thanks for doing this. Before enhancing the solution in future PRs could you use a feature flag and move some of the files to the other place where we store highlight sections?

a8ck3n committed:
Update icons.

a8ck3n commented:
> I noticed the font-family for the amounts is different from what we're using on the other Traffic and Insights pages and also from the screenshots on https://github.com/Automattic/wp-calypso/issues/72045. Are they different on purpose?

That was not on purpose. I didn't notice that as the fonts don't look as thick on my screen as they do in your screenshot. I'll get that fixed up!

a8ck3n committed:
Add missing span for styling.

a8ck3n committed:
Replace instance of GridIcon with Icon.

We need to wrap Icon in a span to allow the Popup to display properly.

a8ck3n committed:
Remove old code paths.

a8ck3n commented:
I've updated the PR to add in the proper icons, update the fonts, and to remove the old code paths. There are still a few bits of feedback unresolved though. Would appreciate your thoughts on those @grzegorz-cp. Thanks!

kangzj commented on client/my-sites/stats/wordads/highlights-section.jsx:
```diff
@@ -0,0 +1,154 @@
+}
+
+function getHighlights( earnings ) {
+	const total = earnings && earnings.total_earnings ? Number( earnings.total_earnings ) : 0;
```
You could use optional chaining to shorten this as `earnings?.total_earnings ? .... ...`.

And several places in the following code could be shortened in the same way.

a8ck3n replied:
Thanks for the reminder. I was meaning to update the logic here. I'll get it updated.

kangzj commented on client/my-sites/stats/wordads/highlights-section.jsx:
```diff
@@ -0,0 +1,154 @@
+			}
+		),
+	};
+	return owed < 100 ? [ notice, limit ] : [ payout ];
```
`owed` is a string and `100` is a number. It's generally not safe to compare strings to numbers - in this case it runs okay but I'd still recommend not doing this.

a8ck3n replied:
I'll fix this up to. I didn't realize I was doing this.

kangzj commented on client/my-sites/stats/wordads/highlights-section.jsx:
```diff
@@ -0,0 +1,154 @@
+		return <h1 className="highlight-cards-heading">{ localizedTitle }</h1>;
+	}
+	return (
+		<h1 className="highlight-cards-heading">
```
Sorry for being picky. The code here are mostly duplication of the early return above. Would it be better only to conditional show the differences?

a8ck3n replied:
I initially had that but it generated an empty DIV in the DOM for some reason. If you clicked on it, it would disappear. Coding it this way resulted in cleaner DOM output. I'll check on this again to see if I was overlooking something.

a8ck3n replied:
I'm still seeing this behaviour in Chrome but not in Firefox. Maybe it's just a quirk with how Chrome is reformatting it's output when browsing the DOM.

kangzj reviewed (commented):
Good job! It works as described.

A tiny styling feedback: On smaller screens the width of the cards are not aligned with the ones on other pages.

<img width="477" alt="image" src="https://user-images.githubusercontent.com/1425433/214157722-79e1e817-6b5b-4932-9995-4f17d73b39d1.png">

<img width="470" alt="image" src="https://user-images.githubusercontent.com/1425433/214157802-e9fba03b-c16c-4418-831a-3bdd09e810db.png">

grzegorz-cp commented on client/my-sites/stats/wordads/highlights-section.jsx:
```diff
@@ -0,0 +1,154 @@
+	const notices = payoutNotices( earningsData );
+	return (
+		<div className="highlight-cards wordads">
+			<HighlightsSectionHeader notices={ notices } />
```
Since those two lines are other functional components, they should be placed in their own files, or they should stay in this one, not other functions.
Were you planning to do it later?

a8ck3n replied:
> Were you planning to do it later?

I wasn't planning to break these out. I was thinking of them more like helper components that didn't necessarily need to be exposed (or exported). I'm of the mind that smaller functions and components are easier to understand when it comes to reading new code (or revisiting old code) and figuring out what's happening.

If it's a best practice we are looking to follow, I guess I'd favour inlining the code but I do feel that hurts readability.

grzegorz-cp reviewed (approved):
I added some suggestions, but nothing is blocking.

a8ck3n commented:
> A tiny styling feedback: On smaller screens the width of the cards are not aligned with the ones on other pages.

Hey @kangzj, thanks for the review!

This was intentional as the strings we are displaying can be much longer. Likewise for the "Outstanding amount" label. I bumped up the minimum width to account for those. See [this commit](https://github.com/Automattic/wp-calypso/pull/72320/commits/2ef43cbde8f0dc92707ffdda471c51d3d79f3104) for my initial thoughts.

Maybe we could introduce a second style (with a smaller `min-width`) and apply the correct class based on the string length. I'll test against the longer labels first because though, as it definitely didn't fit within the previous min.

a8ck3n committed:
Adjust min-width of highlight cards.

Now that we are using the correct proportional font we can fit more in a smaller space.

a8ck3n committed:
Fix cross-type compare.

Track the amount owed as a number for comparison and a string for display.

a8ck3n committed:
Update to use optional chaining.

a8ck3n committed:
Tweak code path for tooltip render.

a8ck3n commented:
Rebasing in the hopes that fixes the (newly) failing E2E tests.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7770919

Thank you @a8ck3n for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7770919).
